backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/ZoomService.cs

```diff
@@ -0,0 +1,111 @@
+using EasyMeets.Core.BLL.Interfaces;
+using EasyMeets.Core.Common.DTO.Credentials;
+using EasyMeets.Core.Common.DTO.Credentials.Zoom;
+using EasyMeets.Core.Common.DTO.Zoom;
+using EasyMeets.Core.Common.Enums;
+using EasyMeets.Core.Common.NamingPolicies;
+using EasyMeets.Core.DAL.Context;
+using EasyMeets.Core.DAL.Entities;
+using Microsoft.AspNetCore.WebUtilities;
+
+namespace EasyMeets.Core.BLL.Services;
+
+public class ZoomService : BaseService, IZoomService
+{
+    private readonly HttpClient _httpClient;
+    private const string TokenUri = "https://zoom.us/oauth/token";
+    private const string BaseApiUri = "https://api.zoom.us/v2";
+    public ZoomService(EasyMeetsCoreContext context, IMapper mapper, HttpClient httpClient) : base(context, mapper)
+    {
+        _httpClient = httpClient;
+    }
+
+    public async Task<CredentialsDto> GetNewCredentials(NewCredentialsRequestDto newCredentialsRequestDto)
+    {
+        var queryString = new Dictionary<string, string?>
+        {
+            { "code", newCredentialsRequestDto.Code },
+            { "grant_type", newCredentialsRequestDto.GrantType },
+            { "redirect_uri", newCredentialsRequestDto.RedirectUri }
+        };
+        return await GetCredentials(queryString);
+    }
+
+    public async Task CreateZoomMeeting(Meeting meeting)
+    {
+        var credentials = meeting.Author.Credentials.FirstOrDefault(cr => cr.Type == CredentialsType.Zoom);
+        if (credentials is null)
+        {
+            throw new KeyNotFoundException("No zoom credentials found for meeting Author");
+        }
+
+        var newMeeting = _mapper.Map<NewZoomMeetingDto>(meeting);
+        
+        using var request = new HttpRequestMessage(HttpMethod.Post, $"{BaseApiUri}/users/me/meetings");
+        var options = new JsonSerializerOptions()
+        {
+            PropertyNamingPolicy = SnakeCaseNamingPolicy.Instance
+        };
+        var token = await GetAccessToken(credentials);
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
+        request.Content = JsonContent.Create(newMeeting, options:options);
+
+        var response = await _httpClient.SendAsync(request);
+
+        response.EnsureSuccessStatusCode();
+
+        var created = await response.Content.ReadFromJsonAsync<ZoomMeetingResponseDto>(options);
+        meeting.MeetingLink = created!.JoinUrl;
+        await _context.SaveChangesAsync();
+    }
+
+    private string GetTokenAuthorization()
+    {
+        var clientId = Environment.GetEnvironmentVariable("ZoomClientId")!;
+        var clientSecret = Environment.GetEnvironmentVariable("ZoomClientSecret")!;
+        var bytes = Encoding.UTF8.GetBytes($"{clientId}:{clientSecret}");
+        return Convert.ToBase64String(bytes);
+    }
+
+    private async Task<string> GetAccessToken(Credentials credentials)
+    {
+        if (DateTimeOffset.UtcNow - credentials.UpdatedAt < TimeSpan.FromSeconds(credentials.LifeCycle * 0.8))
```

Comment:
It would be better to move TimeSpan value to constant

Reply:
Timespan value is dependent on credentials lifecycle, would it be ok if I moved the whole check to a separate method?

Reply:
yes

Reply:
Done